File: backend/app/api/endpoints/votes.py
"""
Unified votes endpoint for all entity types.
"""

import logging
from typing import Dict, List, Optional
from uuid import UUID

from fastapi import APIRouter, Depends, Query

from app.api.dependencies.auth import (
    get_current_admin_user,
    get_current_user,
    get_optional_current_user,
)
from app.api.dependencies.repositories import Repositories, get_repositories
from app.api.schemas.vote import (
    EntityType,
    FlaggedEntitySummary,
    VoteCreate,
    VoteRead,
    VoteSummary,
)
from app.api.services.vote_service import VoteService
from app.api.utils.endpoint_decorators import standard_responses
from app.api.utils.response_patterns import ResponsePatterns
from app.db.dynamo.users import User as DBUser

logger = logging.getLogger(__name__)

# Create router
router = APIRouter()

# Create service
vote_service = VoteService()


@router.get(
    "/count",
    response_model=Dict[str, int],
    responses=standard_responses(
        success_description="Count of votes",
    ),
)
async def count_votes(
    repos: Repositories = Depends(get_repositories),
) -> Dict[str, int]:
    """Get total count of votes."""
    try:
        count = repos.votes.count()
        logger.info(f"Retrieved votes count: {count}")
        return {"count": count}
    except Exception as e:
        logger.error(f"Error counting votes: {str(e)}")
        raise


@router.post(
    "/{entity_type}/{entity_id}",
    response_model=VoteRead,
    responses=standard_responses(
        success_description="Vote created/updated successfully",
        validation_error=True,
        not_found=True,
        conflict=True,
    ),
)
async def vote_on_entity(
    entity_type: EntityType,
    entity_id: UUID,
    vote_data: VoteCreate,
    current_user: DBUser = Depends(get_current_user),
) -> VoteRead:
    """Vote on an entity (car, build list, or global part)."""
    vote = vote_service.vote_on_entity(
        entity_type=entity_type,
        entity_id=entity_id,
        user_id=current_user.id,
        vote_data=vote_data,
        logger=logger,
    )
    return VoteRead.model_validate(vote)


@router.delete(
    "/{entity_type}/{entity_id}",
    responses=standard_responses(
        success_description="Vote removed successfully",
        not_found=True,
    ),
)
async def remove_vote(
    entity_type: EntityType,
    entity_id: UUID,
    current_user: DBUser = Depends(get_current_user),
) -> dict[str, str]:
    """Remove a vote from an entity."""
    removed = vote_service.remove_vote(
        entity_type=entity_type,
        entity_id=entity_id,
        user_id=current_user.id,
        logger=logger,
    )

    if removed:
        return {"message": "Vote removed successfully"}
    else:
        # IN-06: use the centralized error-shape helper instead of raw
        # ``HTTPException`` so the response follows the same {message,
        # error_code, details} contract as the rest of the module.
        ResponsePatterns.raise_not_found("Vote")


@router.get(
    "/{entity_type}/{entity_id}/summary",
    response_model=VoteSummary,
    responses=standard_responses(
        success_description="Vote summary retrieved successfully",
        not_found=True,
    ),
)
async def get_vote_summary(
    entity_type: EntityType,
    entity_id: UUID,
    current_user: Optional[DBUser] = Depends(get_optional_current_user),
) -> VoteSummary:
    """Get vote summary for an entity (public endpoint, authentication optional)."""
    user_id = current_user.id if current_user else None
    return vote_service.get_vote_summary(
        entity_type=entity_type,
        entity_id=entity_id,
        user_id=user_id,
        logger=logger,
    )


@router.get(
    "/admin/flagged/{entity_type}",
    response_model=List[FlaggedEntitySummary],
    responses=standard_responses(
        success_description="Flagged entities retrieved successfully",
        unauthorized=True,
        forbidden=True,
    ),
)
async def get_flagged_entities(
    entity_type: EntityType,
    limit: int = Query(50, ge=1, le=100, description="Maximum number of flagged entities to return"),
    current_user: DBUser = Depends(get_current_admin_user),
) -> List[FlaggedEntitySummary]:
    """Get flagged entities (those with high downvote ratios or reports). Admin only."""
    return vote_service.get_flagged_entities(
        entity_type=entity_type,
        limit=limit,
        logger=logger,
    )

File: frontend/src/api/votes.ts
/**
 * Voting and flagging across car generations, build lists, and parts. One module
 * because every entity type shares the same vote endpoints.
 */

import { apiClient } from './client';
import type {
  FlaggedEntitySummary,
  VoteCreate,
  VoteMutationResult,
  VoteSummary,
} from '../types/Api';

/** Voting and flagging across every votable entity type. */
export const votesApi = {
  voteOnEntity: (
    entityType: 'car_generation' | 'build_list' | 'part',
    entityId: string,
    data: VoteCreate
  ) =>
    apiClient.post<VoteMutationResult>(
      `/votes/${entityType}/${entityId}`,
      data
    ),
  removeVote: (
    entityType: 'car_generation' | 'build_list' | 'part',
    entityId: string
  ) => apiClient.delete<VoteMutationResult>(`/votes/${entityType}/${entityId}`),
  getVoteSummary: (
    entityType: 'car_generation' | 'build_list' | 'part',
    entityId: string
  ) => apiClient.get<VoteSummary>(`/votes/${entityType}/${entityId}/summary`),
  getFlaggedEntities: (
    entityType: 'car_generation' | 'build_list' | 'part',
    limit?: number
  ) =>
    apiClient.get<FlaggedEntitySummary[]>(
      `/votes/admin/flagged/${entityType}`,
      { params: { limit } }
    ),
  countVotes: () => apiClient.get<{ count: number }>('/votes/count'),
};

/** Vote endpoints scoped to parts. */
export const partVotesApi = {
  voteOnPart: (partId: string, data: { vote_type: 'upvote' | 'downvote' }) =>
    votesApi.voteOnEntity('part', partId, {
      vote_type: data.vote_type,
      entity_type: 'part',
      entity_id: partId,
    }),
  removeVote: (partId: string) => votesApi.removeVote('part', partId),
  getVoteSummary: (partId: string) => votesApi.getVoteSummary('part', partId),
  getFlaggedParts: (params?: { limit?: number }) =>
    votesApi.getFlaggedEntities('part', params?.limit),
};

/** Vote endpoints scoped to build lists. */
export const buildListVotesApi = {
  voteOnBuildList: (
    buildListId: string,
    data: { vote_type: 'upvote' | 'downvote' }
  ) =>
    votesApi.voteOnEntity('build_list', buildListId, {
      vote_type: data.vote_type,
      entity_type: 'build_list',
      entity_id: buildListId,
    }),
  removeVote: (buildListId: string) =>
    votesApi.removeVote('build_list', buildListId),
  getVoteSummary: (buildListId: string) =>
    votesApi.getVoteSummary('build_list', buildListId),
  getFlaggedBuildLists: (params?: { limit?: number }) =>
    votesApi.getFlaggedEntities('build_list', params?.limit),
};
